iroha_p2p/src/peer.rs

```diff
@@ -403,13 +476,123 @@ impl Garbage {
+        sync::mpsc::{self, Receiver},
+        time,
+    };
+    use ursa::{encryption::symm::chacha20poly1305::ChaCha20Poly1305, kex::x25519::X25519Sha256};
+
+    use super::*;
+
+    #[derive(iroha_actor::Message, Clone, Debug, Decode, Encode)]
+    struct TestMessage(String);
+
+    fn peer_id() -> PeerId {
+        #[allow(clippy::expect_used)]
+        let port = unique_port::get_unique_free_port().expect("Unable to create free port");
+        let address = format!("127.0.0.1:{}", port);
+        PeerId {
+            address,
+            public_key: None,
+        }
+    }
+
+    fn peer_to<T, K, E>(id: PeerId) -> (Peer<T, K, E>, Receiver<Received<T>>)
+    where
+        T: Encode + Decode + Send + Clone + 'static + Debug,
+        K: KeyExchangeScheme + Send + 'static,
+        E: Encryptor + Send + 'static,
+    {
+        let (sender, receiver) = mpsc::channel(100);
+        (Peer::new_to(id, sender.into()).unwrap(), receiver)
+    }
+
+    fn peer_from<T, K, E>(id: PeerId, stream: TcpStream) -> (Peer<T, K, E>, Receiver<Received<T>>)
+    where
+        T: Encode + Decode + Send + Clone + 'static + Debug,
+        K: KeyExchangeScheme + Send + 'static,
+        E: Encryptor + Send + 'static,
+    {
+        let (sender, receiver) = mpsc::channel(100);
+        (Peer::new_from(id, stream, sender.into()).unwrap(), receiver)
+    }
+
+    async fn peer_ping_pong<T, K, E>() -> (
+        (Addr<Peer<T, K, E>>, Receiver<Received<T>>),
+        (Addr<Peer<T, K, E>>, Receiver<Received<T>>),
+    )
+    where
+        T: Encode + Decode + Send + Clone + 'static + Debug,
+        K: KeyExchangeScheme + Send + 'static,
+        E: Encryptor + Send + 'static,
+    {
+        let id = peer_id();
+        let listener = TcpListener::bind(&id.address).await.unwrap();
+        let jh = tokio::spawn(async move {
+            let (stream, _) = listener.accept().await.unwrap();
+            stream
+        });
+
+        let (to, to_recv) = peer_to(id.clone());
+        let to = to.start().await;
+        tokio::time::sleep(Duration::from_millis(100)).await;
+
+        let (from, from_recv) = peer_from(id, jh.await.unwrap());
+        let from = from.start().await;
+        tokio::time::sleep(Duration::from_millis(100)).await;
+
+        ((to, to_recv), (from, from_recv))
+    }
+
+    #[tokio::test]
+    async fn message_ping_pong() {
+        #[allow(unused)]
+        let telemetry = iroha_logger::init(LoggerConfiguration::default()).unwrap();
+        info!("Starting network tests...");
+
+        let ((to, mut to_recv), (from, mut from_recv)) = time::timeout(
+            Duration::from_secs(1),
+            peer_ping_pong::<TestMessage, X25519Sha256, ChaCha20Poly1305>(),
+        )
+        .await
+        .unwrap();
+
+        let text = String::from("Some data to send to peer");
+
+        // Sending message in one direction...
+        let post = Post {
+            data: TestMessage(text.clone()),
+            id: peer_id(),
```

Comment:
Peer id here is always new. I guess we should change api of peers (remove id from post, as peer always knows to whom he should send it).

Reply:
But that in next prs (i guess with type state probably)

Reply:
But test here doesnt use network...

Reply:
Id will always be the same for same peer

Reply:
> 
> 
> But test here doesnt use network...

I know, but we need some kind of Id here for consistency, and `Post<T>` structure will be used only from `Network` actor.

Reply:
We dont need it though